### PR TITLE
Add support for DATAGRAM frames

### PR DESCRIPTION
**Pull request by mickel8**
This PR adds support for DATAGRAM extension.

### CODE REVIEW

**Comment by qzhuyan** [comment]
Thanks for the PR , I am busy with other things , ping me for review when you think it is ready.

**Comment by mickel8** [comment]
Sure, it should be ready in a few days

**Comment by mickel8** [comment]
@qzhuyan I think that first version of this PR is ready for CR. 

We probably need to add more test cases (e.g. when someone is trying to send datagram bigger than MTU) but let me know what do you think befor I put more effort.

File: c_src/quicer_ctx.h [hunk 1]
```diff
@@ -87,6 +87,15 @@ typedef struct QuicerStreamSendCTX
   ErlNifBinary bin;
 } QuicerStreamSendCTX;
 
+typedef struct QuicerDgramSendCTX
```
Comment (qzhuyan) [inline c_src/quicer_ctx.h:90]: Could it reuse `QuicerStreamSendCTX `?

**Comment by qzhuyan** [comment]
Thanks for the PR, I haven't look into the RFC draft yet, but I guess a test of mixing sending both DATAGRAM and normal QUIC data is needed?

**Comment by mickel8** [comment]
Sorry, quite busy in the last days. I will try to add missing tests in a few days and will ping you again

**Comment by qzhuyan** [comment]
@mickel8 I could merge it if you want?  but some files need to be added with apache license headers as the others.

**Comment by mickel8** [comment]
Sorry for not replying. Please give me two or three more days and I will add missing tests as well as LICENSE files.

**Comment by mickel8** [comment]
@qzhuyan I think I added missing things

File: c_src/quicer_dgram.c [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  QUIC_STATUS Status;
+  if (QUIC_FAILED(Status = MsQuic->DatagramSend(Connection,
+                                                &dgram_send_ctx->Buffer,
+                                                1,
+                                                QUIC_SEND_FLAG_NONE,
+                                                dgram_send_ctx)))
+    {
+      destroy_dgram_send_ctx(dgram_send_ctx);
+      enif_mutex_unlock(c_ctx->lock);
+      return ERROR_TUPLE_3(ATOM_DGRAM_SEND_ERROR, atom_status(Status));
```
Comment (qzhuyan) [inline c_src/quicer_dgram.c:101]: s/atom_status/ATOM_STATUS

This is a recent change on `main` branch

Comment (mickel8) [reply c_src/quicer_dgram.c:101]: Kk, I will fix this today evening

Comment (mickel8) [reply c_src/quicer_dgram.c]: fixed

**Comment by mickel8** [comment]
@qzhuyan  I see that some tests are not passing but I am not sure why. I managed to reproduce error with `transport_shutdown` two times locally but in most cases everything works. Could you take a look at this?

**Comment by qzhuyan** [comment]
Some updates, I am not able to reproduce the issue found in github action on my local machine which is why this PR has been put aside for weeks.

Another interesting finding is the lttng tracing didn't work well in this PR. The tracing is introduced to track down the issue in CI env,  so I will fix that first.

I will clone the branch of this PR and troubleshoot on a checkout branch that I have full control of.

**Comment by qzhuyan** [comment]
> No worries. I thought about some RC that can appear on slower machine (for example during CI pipeline) and I can see that you created draft PR #63. Is it related with my failing tests?

Thanks for the info.

#63 is unrelated, it fixes some race cond. for some API calls while the resources are closed already.

The good news is I managed to reproduce the issue and collect debug logs from github workflow after rebase to the master branch, I can start dig into that now.

File: test/quicer_SUITE.erl [hunk 3]
```diff
@@ -1344,6 +1379,43 @@ ping_pong_server_stm_loop(L, Conn, Stm) ->
...
+      timer:sleep(100),
+      ping_pong_server_dgram(Owner, Config, Port)
+  end.
+
+ping_pong_server_dgram_loop(L, Conn, Stm) ->
+  receive
+    {quic, <<"ping">>, _, _, _, _} ->
+      ct:pal("send pong"),
+      {ok, 4} = quicer:send(Stm, <<"pong">>),
+      ping_pong_server_stm_loop(L, Conn, Stm);
```
Comment (qzhuyan) [inline test/quicer_SUITE.erl:1402]: well, this should be `ping_pong_server_dgram_loop`. I think this is the reason makes the test unstable in github CI.

Comment (mickel8) [reply test/quicer_SUITE.erl]: :man_facepalming: 

Thanks a lot. I have fixed it

**Comment by qzhuyan** [comment]
Could you rebase to latest master?

**Comment by mickel8** [comment]
@qzhuyan  I have merged master to dgram. Is it okay or you prefer rebase?

**Comment by mickel8** [comment]
It looks like something is constantly failing but I cannot see any info that this is because of my code. Should I fix something?

**Comment by qzhuyan** [comment]
The CI env with lttng loggging seems broken, I will check that.

**Comment by qzhuyan** [comment]
I have fixed CI in https://github.com/emqx/quic/pull/65, pls try to rebase/merge. thanks!

**Comment by mickel8** [comment]
Merged, could you run workflow?

**Review by qzhuyan — APPROVED** [review]

**Pull request merged**